feat: Nail Search

pjhul opened the pull request:
## Context

Currently, we use [Algolia Docsearch](https://docsearch.algolia.com/) for search on our website. While this product is great, it unfortunately doesn't suit our needs very well anymore as our website has grown. This means that the current search box gets very little use, even from core team members, and makes everything harder to find.

## Issues with the current search
![Screenshot_20221128_092826](https://user-images.githubusercontent.com/39424187/204342505-b01c1e19-0c67-4b0d-be54-3e2e96fd182d.png)
- Results are too granular - Docsearch lists results based on headings, which can lead to incredibly cluttered results. This works fine for searching through docs that only have a small number of pages like standalone libraries, but for sites like ours that have many-hundreds of pages this is just to granular 
- Data is obtained by scraping the site - Docsearch crawls our site and extracts titles and headings, which excludes a lot of important metadata
- Scraping is only done once per week
- The current box just uses the default styling

## Changes
![Screenshot_20221128_093600](https://user-images.githubusercontent.com/39424187/204343838-84401bad-439f-4e88-a778-4cbc0fd3e330.png)
- All new search box!
- Custom Algolia implementation on the backend
- Can now be opened on any page on the site using either `Ctrl+K` or `/`
- Allows filtering results by type, and when opening the search from specific pages results are pre-filtered
- Page view numbers are factored into ranking (Move viewed generally equals higher ranked)
- Page data is sent to Algolia on every build rather than scraped every week

#### Misc.
- Moved Squeak! schema customizations to the gatsby-source-squeak plugin
- Moved all Gatsby config files to typescript
- Removed some unused Netlify plugins
- Changed the order of imports in PostLayout

## To-do
- [x] Add more granular custom tracking ('Result clicked', 'Result hovered')
- [x] Mobile layout
- [x] Dark mode

corywatilo commented:
Closes #3295

corywatilo commented:
Alright I'm done with styling.

Only blocker I see is:

- [ ] Slugs for results don't populate for the Questions and API sections

And only thing I (still) don't love is having two search bars on the Docs and Handbook index pages because:

- Conflicting styles are ugly and don't follow any of our existing styles (my fault for introducing the round one a while back)
- Repetitive
- It doesn't help train people where to look for the consistent search bar on the content pages

pjhul commented:
@corywatilo I think those points about the search bar are fair - for now I've removed it and we can see how it affects the number of people that are using search. We could also keep playing around with adding it somewhere in the header, which may be the best solution overall.

Fixed that bug with slugs not showing up for questions and the API. Some questions still don't have permalinks (This will be fixed in #4626 so for now I've just linked them to the `/questions` page). I have a couple last small things to do but I think by later today it'll be ready for a final review!

smallbrownbike reviewed (approved):
SOLID. 🎉
